collector/dhcp.go

```diff
@@ -0,0 +1,412 @@
+			nil,
+			nil,
+		),
+		FailoverTransitionsPartnerdownState: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "failover_transitions_partnerdown_state"),
+			"Number of transitions into PARTNER DOWN state (FailoverTransitionsPartnerdownState)",
+			nil,
+			nil,
+		),
+		FailoverTransitionsRecoverState: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "failover_transitions_recover"),
+			"Number of transitions into RECOVER state (FailoverTransitionsRecoverState)",
+			nil,
+			nil,
+		),
+		FailoverBndupdDropped: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "failover_bndupd_dropped"),
+			"Number of DHCP faileover Binding Updates dropped (FailoverBndupdDropped)",
+			nil,
+			nil,
+		),
+	}, nil
+}
+
+// Collect sends the metric values for each metric
+// to the provided prometheus Metric channel.
+func (c *DhcpCollector) Collect(ctx *ScrapeContext, ch chan<- prometheus.Metric) error {
+	if desc, err := c.collect(ctx, ch); err != nil {
+		log.Error("failed collecting dhcp metrics:", desc, err)
+		return err
+	}
+	return nil
+}
+
+// represents perflib metrics from the DHCP Server class.
+// While the name of a number of perflib metrics would indicate a rate is being returned (E.G. Packets Received/sec),
+// perflib instead returns a counter, hence the "Total" suffix in some of the variable names.
+type dhcpPerf struct {
+	PacketsReceivedTotal                             float64 `perflib:"Packets Received/sec"`
+	DuplicatesDroppedTotal                           float64 `perflib:"Duplicates Dropped/sec"`
+	PacketsExpiredTotal                              float64 `perflib:"Packets Expired/sec"`
+	MillisecondsPerPacketAvg                         float64 `perflib:"Milliseconds per packet (Avg)."`
+	ActiveQueueLength                                float64 `perflib:"Active Queue Length"`
+	ConflictCheckQueueLength                         float64 `perflib:"Conflict Check Queue Length"`
+	DiscoversTotal                                   float64 `perflib:"Discovers/sec"`
+	OffersTotal                                      float64 `perflib:"Offers/sec"`
+	RequestsTotal                                    float64 `perflib:"Requests/sec"`
+	InformsTotal                                     float64 `perflib:"Informs/sec"`
+	AcksTotal                                        float64 `perflib:"Acks/sec"`
+	NacksTotal                                       float64 `perflib:"Nacks/sec"`
+	DeclinesTotal                                    float64 `perflib:"Declines/sec"`
+	ReleasesTotal                                    float64 `perflib:"Releases/sec"`
+	OfferQueueLength                                 float64 `perflib:"Offer Queue Length"`
+	DeniedDueToMatch                                 float64 `perflib:"Denied due to match."`
+	DeniedDueToNonmatch                              float64 `perflib:"Denied due to nonmatch."`
```

Comment:
These are almost certainly totals, I'll run this on our DHCP server to confirm the value is increasing over time.

Reply:
I've been running the collector on our DHCP servers for about a week now and haven't seen this metric increment, or change in any way.
Perhaps someone else could run the collector and monitor the metrics for changes?

Reply:
I don't have a Windows DHCP server to test on, unfortunately. I guess we have two options, either leave them out for a first iteration and add them back once someone has a way to validate it, or just go with what seems reasonable and set them to counters and consider it a bug if that turns out to be false.
I'd be fine with the latter, since it seems very likely to be true. What do you think?

Reply:
I agree with the latter, let's treat them as Counters and amend if necessary. I've updated the type in a recent push.